### PR TITLE
iptables/netfilter: rework dependencies

### PR DESCRIPTION
**Pull request by champtar**
iptables-nft is now usable ~and doesn't depends on any kmod-ipt-* anymore~
I tried to rename iptables to iptables-legacy and add PROVIDES but it creates circular dependencies for packages having for example "+iptables +iptables-mod-*"

~To test / migrate your package, replace iptables with iptables-nft, remove iptables-mod-* / kmod-ipt-*, and see what happens :)~

Edit: I reduced the scope of this PR for now as changing all dependencies to not pull the iptables kernel part cause a lot of changes with some chance of regression, so let's make it work, finish packaging all nft modules if some are still missing and have people switch to nft so we don't have to touch iptables :)

### CODE REVIEW

**Comment by champtar** [comment]
filter is working, but nat is broken :(
```
# iptables -t nat -A POSTROUTING -j SNAT --to 1.2.3.4
iptables v1.8.7 (nf_tables): unknown option "--to"
Try `iptables -h' or 'iptables --help' for more information.
```

**Comment by tiagogaspar8** [comment]
Great work!😁

**Comment by champtar** [comment]
This is as broken as before, maybe we need CONFIG_NFT_COMPAT and/or something else, but this is ready for review

**Comment by tiagogaspar8** [comment]
Why is it broken?
I haven't tested it

**Comment by ptpt52** [comment]
To follow this thread.

File: package/network/utils/iptables/Makefile [hunk 1]
```diff
@@ -454,17 +454,17 @@ endef
...
-  iptables-nft
-  iptables-nft-restore
-  iptables-nft-save
-  iptables-translate
-  iptables-restore-translate
+  ip6tables-nft
+  ip6tables-nft-restore
+  ip6tables-nft-save
+  ip6tables-translate
+  ip6tables-restore-translate
```
Comment (aparcar) [inline package/network/utils/iptables/Makefile:483]: Am I wrong or is this an untreated fixup?

File: package/network/utils/iptables/Makefile [hunk 2]
```diff
@@ -598,6 +598,7 @@ endef
 define Package/iptables/install
 	$(INSTALL_DIR) $(1)/usr/sbin
 	$(CP) $(PKG_INSTALL_DIR)/usr/sbin/xtables-legacy-multi $(1)/usr/sbin/
+	$(CP) $(PKG_INSTALL_DIR)/usr/sbin/iptables-legacy{,-restore,-save} $(1)/usr/sbin/
```
Comment (aparcar) [inline package/network/utils/iptables/Makefile:617]: Please add a commit message describing what you're doing. I'm guessing building `iptables` generates symlinks already, right?

**Review by aparcar — CHANGES_REQUESTED** [review]

**Comment by adschm** [comment]
Extensive commit descriptions would be a good idea indeed. Apart from that, some patches look like the should have a "Fixes:" in the commit message as well.

**Comment by champtar** [comment]
I'll improve commits messages tonight, not sure if there is an open issue for what I'm fixing

File: package/network/utils/iptables/Makefile (hunk 1 above)
Comment (champtar) [reply package/network/utils/iptables/Makefile:483]: moved to another commit

File: package/network/utils/iptables/Makefile (hunk 2 above)
Comment (champtar) [reply package/network/utils/iptables/Makefile:617]: Improved a bit the commit message
```
 iptables: add ip{,6}tables-legacy{,-restore,-save} symlinks

Now that we can have both legacy and nft iptables variants
installed at the same time, install the legacy symlinks
```

Comment (champtar) [reply package/network/utils/iptables/Makefile:617]: yup iptables build generate the symlinks

**Comment by champtar** [comment]
To make iptables-nft really usable, we would need to have all xt_* modules install-able without any ipt_* / iptable_filter / ...
Please have a look at what I've already done

File: package/network/utils/iptables/Makefile [hunk 3]
```diff
@@ -611,7 +626,6 @@ endef
 
 define Package/ip6tables/install
 	$(INSTALL_DIR) $(1)/usr/sbin
-	$(CP) $(PKG_INSTALL_DIR)/usr/sbin/ip6tables{,-restore,-save} $(1)/usr/sbin/
```
Comment (hauke) [inline package/network/utils/iptables/Makefile:614]: Please mention in the comit message that this just installs symlinks and that the ALTERNATIVES attribute does more or less the same.

**Comment by hauke** [comment]
The code changes are looking fine to me and the split into multiple commits is also nice.
Could you please add a better description to the commit messages why you are doiug wand why.

**Comment by nasbdh9** [comment]
```
root@OpenWrt:~# lsmod | grep tproxy
nf_tproxy_ipv4         16384  1 xt_TPROXY
nf_tproxy_ipv6         16384  1 xt_TPROXY
root@OpenWrt:~#
root@OpenWrt:~# nft add rule v2ray prerouting meta l4proto {tcp, udp} mark set 1 tproxy to 127.0.0.1:12345 accept
Error: Could not process rule: No such file or directory
add rule v2ray prerouting meta l4proto {tcp, udp} mark set 1 tproxy to 127.0.0.1:12345 accept
                                                             ^^^^^^^^^^^^^^^^^^^^^^^^^
```
When I try to use tproxy, the system prompts that the file is missing, and nf_tproxy is already loaded. Is there anything else missing?

**Comment by champtar** [comment]
Sorry I've been busy, I've switched to draft again as this is not ready. I think I'll do less change in the end as there is always a corner case that breaks, more details to follow.

**Comment by LGA1150** [comment]
> ```
> root@OpenWrt:~# lsmod | grep tproxy
> nf_tproxy_ipv4         16384  1 xt_TPROXY
> nf_tproxy_ipv6         16384  1 xt_TPROXY
> root@OpenWrt:~#
> root@OpenWrt:~# nft add rule v2ray prerouting meta l4proto {tcp, udp} mark set 1 tproxy to 127.0.0.1:12345 accept
> Error: Could not process rule: No such file or directory
> add rule v2ray prerouting meta l4proto {tcp, udp} mark set 1 tproxy to 127.0.0.1:12345 accept
>                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^
> ```
> 
> When I try to use tproxy, the system prompts that the file is missing, and nf_tproxy is already loaded. Is there anything else missing?

You need `nft_tproxy` for that. `xt_TPROXY` is for iptables or `nft_compat`.

File: package/network/utils/iptables/Makefile (hunk 3 above)
Comment (champtar) [reply package/network/utils/iptables/Makefile:614]: done

**Comment by champtar** [comment]
The must have changes of this PR are adding nft-compat and using alternatives
Reworking the whole maze of dependencies is painfull for not much gain, let's just have people switch to nft :)

**Comment by aparcar** [comment]
aa63c7aa279a0d25891383e805e9bb796da6f90c I'd prefer `netfilter` instead of `netfilter.mk`, looking at the other files and some earlier commits, that seems to be the way. 

84fa9cbc3109c1b1a1142eeaa62c0b6c56236370 please add a commit message
9a703c6ef11ac7f9f0271da958c30bcfe0caa595 please add a commit message

**Comment by hnyman** [comment]
@champtar 

Regarding your advice in the sqm-scripts PR:

> You actually need to keep the iptables-mod-* dependencies 

Both iptables-mod-ipopt  and iptables-mod-conntrack-extra seem to depend on the userspace iptables package. They can't be selected if iptables is not selected. Is that still necessary?   (your current PR 5004 (commit 32f4cf4) is applied)

That seems to hold true for all the iptables-mod-something packages due to the dependency in the generic module macro
https://github.com/openwrt/openwrt/blob/32f4cf49dcad07d44e7f6cd4a0a215c42f7e5617/package/network/utils/iptables/Makefile#L42-L45

**Comment by nasbdh9** [comment]
same question, after selecting `nft_compat` and `iptables-nft`, how should the original `iptables-mod-*` be handled?

**Comment by champtar** [comment]
@aparcar fixed
@hnyman @nasbdh9 in the end I'm not fixing all the dependencies as I wanted to do initialy, so yes for now switching to iptables-nft you will still have iptables present (but not used). the iptables-legacy modules will be loaded showing a warning in iptables-nft-save but I don't think this is an issue.

What needs to be done/could be done:
- add something like PKG_CONFIG_DEPENDS:=IPT_BUILTIN to iptables Makefile
- split the kmod in smaller groups in netfilter.mk, i've tried but I add some weird results, might be because I was not recompiling iptables ...
- rename from kmod-ipt to kmod-xt
- add iptables-mod-core and remove iptables BUILTIN_MODULES/IPT_BUILTIN entirely so that nft correctly display rules created by iptables-nft

Having this PR merged still allows people to really use iptables-nft, to see if "it works (tm)"

**Comment by hnyman** [comment]
Great that you have been working on this. 
Hopefully the current (or some polished) version gets merged soon, so that wider testing can be done easier.

I applied the current version yesterday in my build, and seemed to work. 

However, I had to manually select iptables in order to see the iptables-mod-xxx packages. If I did not select iptables, e.g. SQM was not selectable at all, as it depends not only on iptables-nft but with iptables-mod-ipopt modules.  My intuition is that in the current nftables/iptables-nft world it would be better to have that iptables-mod recipe to have inclusive "+iptables" dependency instead of "iptables", but I haven't tested it.   (Other approach might be to have iptables-nft to depend on +iptables, but that sounds worse)

**Comment by hnyman** [comment]
I built today with the current  b9a73ca  plus the change of "+iptables" in the iptables-mod recipe at
https://github.com/openwrt/openwrt/blob/b9a73ca/package/network/utils/iptables/Makefile#L42-L45

That helped in making packages depending on iptables-mod-xxx (like SQM) able to select the iptables-mod-xxxxx package (e.g. iptables-mod-ipopt for SQM) without adding a iptables dependency to the package, or manually adding iptables to my build.

As iptables-mod-xxx packages currently require iptables to have been selected earlier, I see no harm in changing the  iptables-mod recipe to be selective.

**Comment by champtar** [comment]
@hnyman can you add the +iptables when merging this PR ? This week is a bit complicated for me (traveling).

**Comment by hnyman** [comment]
Sorry, I do not have commit rights in the main repo...

**Comment by champtar** [comment]
> Sorry, I do not have commit rights in the main repo...

Don't know why I was sure you were a core maintainer. I'll add a commit and do minimal build test when I get a chance then.

**Comment by hauke** [comment]
Thank you for your patches, I applied them to master in a5c8811c04799a601447a5b0cd21d39e00c0ae0e, 4e7ad159043f3c75eaf4d81a2dc089d7babba07e, 1ebb8e3b6be8ed070dcb466fc8d1631e5453df72, b0bd6599e840c443d8ccb8759315e4dd006fd6aa, 3a5df36cf694ca821ad5486ff360969bd3492aaa and 0e32c6baf37688e8cefed5718afe13ec18f75233

I added the + to the iptables dependency.

**Pull request closed without merging**